Upgrade Bid Adjustments

AlexBVolcy opened the pull request:
Addresses this feature: https://github.com/prebid/prebid-server/issues/2085

The goal of this feature is to add support for bid adjustments that works differently and is more complex than the existing `bidadjustmentfactors`. 

This feature allows `bidadjustments` to be defined on the request at the path `ext.prebid.bidadjustments`. Within this definition, `bidadjustments` will specify the mediatype, the bidder name, the dealID, and finally an adjustment array.

This adjustment array will contain information that'll be used to the adjust the bid price of a particular bid that matches the given mediatype, bidder name, and dealID. 

An entry in the adjustment array will include the following information: an `adjtype`, a `value`, and sometimes a `currency`

`Adjtype` refers to the type of adjustment that'll be occurring. A `mulitplier` adjtype will result in the bid price being multiplied by the given value. A `static` adjtype will result in the bid price being overwritten by the given value (after currency conversions). A `cpm` adjtype, will result in the given value being subtracted from the bid price.

AlexBVolcy commented on endpoints/openrtb2/auction.go:
```diff
@@ -1549,6 +1549,17 @@ func validateRequestExt(req *openrtb_ext.RequestWrapper) []error {
+		if prebid.Debug {
+			errs = append(errs, &errortypes.Warning{
+				WarningCode: errortypes.BidAdjustmentWarningCode,
+				Message:     "Bid Adjustment From Request Was Invalid",
```
Good point to bring up! I spoke about this with @hhhjort and @bsardo and we settled on doing the validation twice for the time being

AlexBVolcy commented on exchange/bidder.go:
```diff
@@ -680,3 +686,27 @@ func compressToGZIP(requestBody []byte) []byte {
 	return b.Bytes()
 }
+
+// TODO: If adjarray == nil
```
Good catch, will update this

AlexBVolcy commented on exchange/bidder.go:
```diff
@@ -680,3 +686,27 @@ func compressToGZIP(requestBody []byte) []byte {
+			if err != nil {
+				return originalBidPrice
+			}
+			bidPrice = bidPrice - convertedVal
```
Good point! Will address this

AlexBVolcy commented on exchange/bidder.go:
```diff
@@ -331,10 +331,16 @@ func (bidder *bidderAdapter) requestBid(ctx context.Context, bidderRequest Bidde
 						if bidResponse.Bids[i].Bid != nil {
 							originalBidCpm = bidResponse.Bids[i].Bid.Price
 							bidResponse.Bids[i].Bid.Price = bidResponse.Bids[i].Bid.Price * adjustmentFactor * conversionRate
+							bidResponse.Bids[i].Bid.Price = applyAdjustmentArray(adjArray, bidResponse.Bids[i].Bid.Price, bidResponse.Currency, reqInfo)
```
I like that idea! I think I'd want to bring it up during team time first before going through with this :)

AlexBVolcy replied:
I realized after thinking about this, and I think we need to keep these two functions separate. The `GetAdjustmentArray` function is needed on it's own to satisfy the requirement of: "Any module that needs to reverse the adjustment (e.g. floors), will need a way to get the relevant adjustment array and they can walk the adjustments backwards. This implies that there should be a function similar to the Floors getFloor() function that allows callers to supply dimension values and get the adjustments array."

VeronikaSolovei9 replied:
You can still have `GetAdjustmentArray` as a separate function and call it first in the new function that will consolidate 
`GetAdjustmentArray` and `applyAdjustmentArray`. It seems like a logical way to do this, it will minimize functions calls from `requestBid`, it's already complicated. And also `adjArray` is only used in `applyAdjustmentArray`.

AlexBVolcy commented on openrtb_ext/request.go:
```diff
@@ -312,3 +344,116 @@ func (m ExtMultiBid) String() string {
+	case AdjTypeCpm:
+		if adjustment.Currency == nil || adjustment.Value < 0 || adjustment.Value > math.MaxFloat64 {
+			return false
+		}
```
Good point! I'll bring this up during team time as well to see what others think

AlexBVolcy commented on openrtb_ext/request.go:
```diff
@@ -312,3 +344,116 @@ func (m ExtMultiBid) String() string {
+			for _, adjustment := range bidAdjMap[bidderName][dealId] {
+				if !validateAdjustment(adjustment) {
+					return false
+				}
```
I can't return immediately here because there could still be more adjustments that need to be validated. So I'd only return immediately if I get an invalid adjustment.

VeronikaSolovei9 reviewed (commented):
Thank you for addressing comments, looks good! Added some more comments and I also missed the case with decimal places while testing it previous time.

VeronikaSolovei9 commented on exchange/exchange.go:
```diff
@@ -277,6 +277,14 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		r.FirstPartyData = resolvedFPD
 	}
 
+	mergedBidAdj, err := mergeBidAdjustments(r.BidRequestWrapper, r.Account.BidAdjustments)
+	if err != nil {
+		return nil, err
+	}
+	if valid := mergedBidAdj.ValidateBidAdjustments(); !valid {
+		mergedBidAdj = nil
+	}
+
```
By "combining" I don't mean to literally merge the code from both of this functions. You can still call them separately, but from another consolidates function. Hope it makes sense.
Something like this (or maybe you will come up with better solution while coding it):
```go
func ProcessBidAdjustments(...){
     mergeBidAdjustments()
    validateBidAdjustments()
}
```

AlexBVolcy replied:
I like this idea, I'll update to combine them!

VeronikaSolovei9 commented:
About `bidPrice = bidPrice - convertedVal`
Round the final number to 4 decimal digits first and then check if it's `if bidPrice <= 0 {`. 
In my example result after subtraction is `0.000009999999999621423` which rounds to `0.0000` and it leads to bid rejection and this message in debug logs: 

```json
"errors": {
            "appnexus": [
                {
                    "code": 999,
                    "message": "Bid \"9109836616608581859\" does not contain positive 'price' which is required since there is no deal set for this bid"
                }
            ]
        },
```

Also `roundTo := math.Pow(10, amountOfDecimalPlaces)` looks like a good candidate to be a constant: 
`const roundTo = 10000`

Here is my final end of this function:

```go
...
rounded := math.Round(bidPrice*roundTo) / roundTo // Returns Bid Price rounded to 4 decimal places

if rounded <= 0 {
	return originalBidPrice
}
return rounded
```

VeronikaSolovei9 commented:
Do you think it makes sense to move all new functions related to bid adjustments from bidder.go and exchange.go to a separate file in `exchange` directory?

AlexBVolcy commented:
@VeronikaSolovei9 just pushed a commit that moves adjustment functions that were present in bidder.go and exchange.go into it's own file in exchange directory. Also moved tests tied to those functions into their own file as well. I also updated the rounding logic!

AlexBVolcy commented:
Getting this failure on the new test file I added:

exchange/bidadjustments_test.go:6:2: no required module provides package github.com/prebid/openrtb/v17/openrtb2; to add it: go get github.com/prebid/openrtb/v17/openrtb2

Just want to highlight this for tomorrow's team time

VeronikaSolovei9 commented:
It feels like you need to merge your branch with latest master. Even if there are no conflicts. 
There are should not be any usages of `github.com/prebid/openrtb/v17/openrtb2`

VeronikaSolovei9 commented:
Use `"github.com/prebid/openrtb/v19/openrtb2"` in bidadjustments_test.go

VeronikaSolovei9 commented on exchange/bidadjustments_test.go:
```diff
@@ -0,0 +1,405 @@
+import (
+	"testing"
+
+	"github.com/prebid/openrtb/v17/openrtb2"
```
Use `"github.com/prebid/openrtb/v19/openrtb2"` instead

VeronikaSolovei9 commented:
Thank you for addressing comments! Changes look good :)

AlexBVolcy commented on endpoints/openrtb2/sample-requests/valid-whole/exemplary/bidadjustments-static.json:
```diff
@@ -0,0 +1,92 @@
+            {
+              "id": "appnexus-bid",
+              "impid": "some-impression-id",
+              "price": 2.5,
```
Yes I believe you're right, it should just be 5.0.

Upon re-reviewing requirements, I think I was wrongly handling currency conversion for the static and cpm bid adjustments.

For static: 

> if adjtype is "static", overwrite the bid response price to the currency-adjusted static value and the currency to the supplied static currency.

I take this as meaning that the bid adjustment value given is already in currency-adjusted to whatever currency is provided, and that I should overwrite the bid response price to be exactly what the bid adjustment value is. And then I should also update the bid response currency to be the currency provided with the bid adjustment. And the commit I'm about to push handles this case as I've just described it.

For CPM:

> if adjtype is specified and "cpm", the first task is to convert the adjustment to the bid currency. Then subtract that value from the price. e.g if the bid price is $2.00 USD, and the cpm is 0.01 EUR, the adjusted price would be $2.00- $0.011=1.989. Round to 4 decimal places.

Currently, I was converting the provided cpm value from the bid's currency to the bid adjustment's currency. But what I need to be doing is the opposite, which is clearly described in the example above. My next commit will address this as well.

AlexBVolcy commented on endpoints/openrtb2/sample-requests/valid-whole/exemplary/bidadjustments-static.json:
```diff
@@ -0,0 +1,92 @@
+        }
+      ],
+      "bidid":"test bid id",
+      "cur":"USD",
```
Yes this should be "EUR", my comment above details how I'll address this.

bsardo commented on exchange/exchange.go:
```diff
@@ -290,6 +290,11 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 		r.FirstPartyData = resolvedFPD
 	}
 
+	mergedBidAdj, err := processBidAdjustments(r.BidRequestWrapper, r.Account.BidAdjustments)
```
Alex and I chatted offline. No action is required here. Yes we will halt processing if the merge fails but this technically should never happen.

AlexBVolcy commented on openrtb_ext/request.go:
```diff
@@ -312,3 +343,106 @@ func (m ExtMultiBid) String() string {
+	return nil
+}
+
+func (bidAdjustments *ExtRequestPrebidBidAdjustments) ValidateBidAdjustments() bool {
```
Discussed offline, moving this validation logic to `openrtb2/auction.go`, creates an import cycle. We decided to move all bid adjustment logic, including validation, into it's own package.

SyntaxNode commented on bidadjustments/validate.go:
```diff
@@ -0,0 +1,61 @@
+		if adjustment.Value >= 0 && adjustment.Value < 100 {
+			return true
+		}
+		adjustment.Currency = ""
```
Why is the currency set to empty after the validation?

AlexBVolcy replied:
I initially did this because of the line from the issue that says during validation, if adjtype is "multiplier", ignore currency. 

But reading this again, ignoring currency is accomplished with how the `apply()`method works, where if the adjtype is multiplier, I don't utilize currency at all in the bid price calculation.

So i think this line could be removed

SyntaxNode commented on endpoints/openrtb2/sample-requests/valid-whole/exemplary/bidadjustments-cpm.json:
Please use a formatter for new json tests. The one built into VS Code is a good choice.

SyntaxNode commented on openrtb_ext/request.go:
```diff
@@ -149,6 +158,28 @@ type ExtRequestPrebidCacheVAST struct {
+// Overall: BidderName maps to a DealID that maps to an Adjustments Array
+type MediaType struct {
+	Banner   map[string]map[string][]Adjustments `json:"banner,omitempty"`
+	Video    map[string]map[string][]Adjustments `json:"video,omitempty"`
```
It's hard to understand `map[string]map[string][]Adjustments` and the comment is evidence of it. Could you use type aliases? Such as...

```
type AdjusmentsByDealID map[string][]Adjustments

map[BidderName]AdjusmentsByDealID
```

SyntaxNode commented on openrtb_ext/request.go:
```diff
@@ -149,6 +150,28 @@ type ExtRequestPrebidCacheVAST struct {
+	WildCard map[string]map[string][]Adjustments `json:"*,omitempty"`
+}
+
+type Adjustments struct {
```
It's odd to see the plural `Adjustments` used as an item in an array/slice. Consider naming this type to the singular `Adjustment`.

SyntaxNode commented on openrtb_ext/request.go:
```diff
@@ -149,6 +150,28 @@ type ExtRequestPrebidCacheVAST struct {
+}
+
+type Adjustments struct {
+	AdjType  string  `json:"adjtype,omitempty"`
```
Consider calling this field name `Type` since this is already within an adjustment object.

SyntaxNode reviewed (commented):
Looks great. I have nitpicks, but nothing major.

AlexBVolcy commented on bidadjustments/get.go:
```diff
@@ -0,0 +1,45 @@
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+func GetAndApplyAdjustmentArray(ruleToAdjustments map[string][]openrtb_ext.Adjustments, bidInfo *adapters.TypedBid, bidderName openrtb_ext.BidderName, currency string, reqInfo *adapters.ExtraRequestInfo) (float64, string) {
```
The `Process` method merges the bid adjustments from the request and account, and it validates this merged info. 

The `GetAndApplyAdjustmentArray` method, gets the relevant adjustment array for the given bid info, and applies that adjustment array to the bid price.

AlexBVolcy commented on bidadjustments/bidadjustments.go:
```diff
@@ -0,0 +1,146 @@
+func applyAdjustmentArray(adjArray []openrtb_ext.Adjustments, bidPrice float64, currency string, reqInfo *adapters.ExtraRequestInfo) (float64, string) {
+	if adjArray == nil {
+		return bidPrice, currency
+	}
```
I just want to check that if there are no adjustments to apply, to return the original bid price and currency. So i can add a length check here as well I think.

AlexBVolcy commented on bidadjustment/get.go:
```diff
@@ -0,0 +1,47 @@
+// given the bid info, we create the same format of combinations that's present in the key of the ruleToAdjustments map
+// the slice is ordered by priority from highest to lowest, as soon as we find a match, we return that slice
+func get(ruleToAdjustments map[string][]openrtb_ext.Adjustment, bidType, bidderName, dealID string) []openrtb_ext.Adjustment {
+	priorityRules := [maxNumOfCombos]string{}
```
Yes these are the same

VeronikaSolovei9 commented on endpoints/openrtb2/sample-requests/valid-whole/exemplary/bidadjustments-cpm.json:
```diff
@@ -0,0 +1,105 @@
+{
```
Wow! Changes and local testing look good. 
Just an idea: do you think it makes sense to move bidadjustments json tests to a separate directory `sample-requests/bidadjustments`?

bsardo commented on bidadjustment/apply_test.go:
```diff
@@ -0,0 +1,447 @@
+	}
+}
+
+func TestGet(t *testing.T) {
```
Can you add one more test case here where none of the priority rules match the bid response?

bsardo reviewed (commented):
Looks really good. Mostly nitpicky comments.

Also can you add doc comments to all public functions at a minimum and use your own discretion on whether to add comments for private functions?

AlexBVolcy commented on bidadjustment/apply.go:
```diff
@@ -0,0 +1,88 @@
+	}
+	roundedBidPrice := math.Round(bidPrice*pricePrecision) / pricePrecision
+
+	if roundedBidPrice <= 0 {
+		return originalBidPrice, originalCurrency
+	}
```
It came as a response to this comment https://github.com/prebid/prebid-server/pull/2678#issuecomment-1509555774

AlexBVolcy commented on bidadjustment/apply.go:
```diff
@@ -0,0 +1,88 @@
+		return bidPrice, currency
+	}
+	originalBidPrice := bidPrice
+	originalCurrency := currency
```
The only time currency is altered is when the adjtype is static, and in validation we allow a static value to be equal to zero. So i think that would be a scenario in which the bid price could be equal to 0 after adjustments, and we'd want to return the original bid currency and price.

bsardo commented on bidadjustment/build_rules.go:
```diff
@@ -0,0 +1,93 @@
+	if err != nil {
+		return nil, err
+	}
+	if !Validate(mergedBidAdj) {
```
Discussed offline. We will stick with a warning but will need to adjust the warning message to indicate the issue is in the account instead of the request. We'll also need to append the error to the errors slice if the error returned by Merge is non-fatal.

VeronikaSolovei9 reviewed (approved):
LGTM, great work!

bsardo commented on endpoints/openrtb2/auction.go:
```diff
@@ -1554,6 +1555,15 @@ func validateRequestExt(req *openrtb_ext.RequestWrapper) []error {
 
+	if !bidadjustment.Validate(prebid.BidAdjustments) {
+		prebid.BidAdjustments = nil
+		reqExt.SetPrebid(prebid)
```
No action required here. We shouldn't need to call `RebuildRequest` at this point. The call to `SetPrebid` is modifying the request ext object on the wrapper and marking it as dirty which is sufficient. The code that follows operates off of the request ext object and does not need access to a marshaled ext.  In `HoldAuction`, we rebuild the request prior to cleaning and splitting the request by bidder to ensure that each bidder uses a copy of the mutated request including any modifications made to the request ext through calls to methods like `SetPrebid`.

VeronikaSolovei9 reviewed (approved):
LGTM!
